test/validation/api/crypto/odp_crypto_test_inp.c

```diff
@@ -1095,12 +1132,187 @@ static void test_capability(void)
 	CU_ASSERT((~capa.auths.all_bits & capa.hw_auths.all_bits) == 0);
 }
 
+/*
+ * Create a test reference, which can be used in tests where the hash
+ * result is within auth_range.
+ *
+ * The ciphertext packet and the hash are calculated using an encode
+ * operation with hash_result_offset outside the auth_range and by
+ * copying the hash in the ciphertext packet.
+ */
+static void create_hash_test_reference(odp_auth_alg_t auth,
+				       const odp_crypto_auth_capability_t *capa,
+				       crypto_test_reference_t *ref,
+				       uint32_t digest_offset,
+				       uint8_t digest_fill_byte)
+{
+	odp_crypto_session_t session;
+	uint32_t n;
+	int rc;
+	odp_packet_t pkt;
+	odp_bool_t ok;
+	const uint32_t auth_bytes = 100;
+	uint32_t enc_digest_offset = auth_bytes;
+	odp_packet_data_range_t cipher_range = {.offset = 0, .length = 0};
+	odp_packet_data_range_t auth_range = {.offset = 0};
+
+	ref->auth_key_length = capa->key_len;
+	ref->auth_iv_length = capa->iv_len;
+	ref->digest_length = capa->digest_len;
+	if (capa->bit_mode)
+		ref->length = auth_bytes * 8;
+	else
+		ref->length = auth_bytes;
+	auth_range.length = ref->length;
+
+	for (n = 0; n < ref->auth_key_length; n++)
+		ref->auth_key[n] = n;
+	for (n = 0; n < ref->auth_iv_length; n++)
+		ref->auth_iv[n] = n;
+	for (n = 0; n < ref->length; n++)
+		ref->plaintext[n] = n;
+
+	memset(ref->plaintext + digest_offset, digest_fill_byte, ref->digest_length);
+
+	pkt = odp_packet_alloc(suite_context.pool, auth_bytes + ref->digest_length);
+	CU_ASSERT_FATAL(pkt != ODP_PACKET_INVALID);
+
+	rc = odp_packet_copy_from_mem(pkt, 0, auth_bytes, ref->plaintext);
+	CU_ASSERT(rc == 0);
+
+	session = session_create(ODP_CRYPTO_OP_ENCODE, ODP_CIPHER_ALG_NULL,
+				 auth, ref, PACKET_IV, HASH_NO_OVERLAP);
+
+	if (crypto_op(pkt, &ok, session,
+		      ref->cipher_iv, ref->auth_iv,
+		      &cipher_range, &auth_range,
+		      ref->aad, enc_digest_offset))
+		return;
```

Comment:
Should there be an assert in the failure branch as well?

Reply:
There are more specific asserts down in alg_op() and alg_packet_op() so an assert has already fired in the failure case here.